pkg/querier/stats_renderer_test.go

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querier
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/grafana/dskit/user"
+	"github.com/grafana/regexp"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/promslog"
+	"github.com/prometheus/common/route"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/promqltest"
+	v1 "github.com/prometheus/prometheus/web/api/v1"
+	"github.com/stretchr/testify/require"
+
+	mimir_stats "github.com/grafana/mimir/pkg/querier/stats"
+)
+
+// Mimir doesn't support Prometheus' UTF-8 metric/label name scheme yet.
+func init() {
+	model.NameValidationScheme = model.LegacyValidation
+}
+
+func TestStatsRenderer(t *testing.T) {
+
+	testCases := map[string]struct {
+		expr                 string
+		expectedTotalSamples uint64
+	}{
+		"zero_series": {
+			expr:                 `zero_series{}`,
+			expectedTotalSamples: 0,
+		},
+		"dense_series": {
+			expr:                 `dense_series{}`,
+			expectedTotalSamples: 11,
+		},
+		"start_series": {
+			expr:                 `start_series{}`,
+			expectedTotalSamples: 6,
+		},
+		"end_series": {
+			expr:                 `end_series{}`,
+			expectedTotalSamples: 6,
+		},
+		"sparse_series": {
+			expr:                 `sparse_series{}`,
+			expectedTotalSamples: 5 + 4,
```

Comment:
shouldn't this be 2 because there are only two samples

```
			sparse_series 0 _ _ _ _ _ _ 7 _ _ _
```

Reply:
i wonder if it's not because of the [lookback period of 5 minutes](https://prometheus.io/docs/prometheus/latest/querying/basics/#staleness). Can you quickly verify that it is? It's configurable on the promQL engine config via `LookbackDelta`

Reply:
Good call, I tried different `LookbackDelta` values and it's consistent with the samples returned.